[WIP] Implement row polymorphism

manmatteo opened the pull request:
Implement row polymorphism as described in Giuseppe Castagna and Loïc Peyrot: [Polymorphic Records for Dynamic Languages](https://www.irif.fr/~gc/papers/oopsla25.pdf)

manmatteo committed:
Start adding syntax for row variables

E-Sh4rk commented:
Seems nice! 😄 

- In `transform.ml`, I see `(* TODO compute a better approximation *)`. Currently the function `simplify` is supposed to preserve semantic equivalence (it is not approximating). I dont know if we can do anything smart when we have row variables that would preserve semantic equivalence. If we have useful ways to over-approximate, we can add another function `approximate` maybe?

- `RowVar of Var.t` so we are using the same kind of variables than for regular type variables. Why not. But they should be treated differently in substitutions though. So maybe it would be less confusing to have another type for row variables ?

- I dont remember how subtyping works for `is_empty`, is it enough to treat them as an open record? (this is the behavior currently implemented I think, with the line 91 `| None, (Open | RowVar _) -> OTy.any ()`)

- Missing: substitution of row variables, tallying, anything else?

Tchou commented:
Hi,

[summary of a discussion with @manmatteo]

for reference, the subtyping algorithm for polymorphic records is in:
[https://www.irif.fr/~gc/papers/oopsla25.pdf](https://www.irif.fr/~gc/papers/oopsla25.pdf),
Page 14 function **Φ**. It's unclear to me whether the it is possible to encode the algorithm as tuples (I think so, but I would have to work a bit on that).

Even though the are not strictly needed for subtyping, I think it is worth think a bit of a correct representation of record types to implement tallying. From my understanding:
-`OTy` (optional types), that is `t∨⊥` now become field types 𝜏 (which can be any combination of `t`, `⊥` and `𝜃`, a field variable that can be instantiated with than field type p9 of the paper). 
- The flag of a record is now `𝜍` which ranges over `..`, `𝜖` or `𝜌` which are row variables, that can be instantiated by rows, which are records + a finite set of label that do not occur in them.

The subtyping algorithm looks straightforward. So I would recommend to:
- [maybe ?] add a meta-data inside variables to indicate whether they are type/field/row variables. This could be used for defensive programming to ensure that we don't mix substitutions for various kinds of variables, and for debugging purposes. We could also use phatom types to enforce the kind at the type level but I think it will make the API needlessly verbose. So a simple solution such as:
```ocaml
type kind = Type | Field | Row
type t = int * string * kind

let mk ?(kind=Type) name = (... new counter, name, kind)

```
- implement the subtyping algorithm as in the paper and test it.
- Then see if it can be encoded as tuples.

@E-Sh4rk what do you think ?

E-Sh4rk commented:
I agree with all of that :)

About variables, I agree that it may be more convenient to have only one type supporting Type, Field and Row variables.
But what about type substitutions? Row substitutions, for instance, are quite different from type substitutions.
@Tchou Do you also think that the type for substitutions should be extended to handle all kind of substitutions?
Or should we define a distinct type for row substitutions? (the former seems to be more in agreement with the fact to represent all kind of variables using the same type).

Tchou committed:
Make the `FieldTy.t` type abstracts

- add constructor and destructor functions to build/extract
  the pair `(t, bool)` (for now). These will be updated to deal with
  field variables.
- Propagate some more type equalities in modules signatures and functor
  definitions. Since the type `FieldTy.t` is now abstract, OCaml cannot
  know that two instanciations of `FieldTy.(N1)` and `FieldTy(N2)` are
  the same, with only the constraint `N1.t = N2.t`.

Tchou committed:
Put back the documentation for `Ty.F`

Tchou committed:
Do not restrict the signature of `FieldTy` in `records.ml`, this way
the subtyping algorithm as access to the underlying representation

Tchou committed:
Implement `hash`, `equal` and `compare` in`Node` outside of `Node`.

By implementing these three function outside of `Node` (in the auxiliary
module `AnyEmpty` we allow their use early, before the recursive module
`Node` is initialized). This is necessary to define the leafs of `Bdd`
as toplevel values (since they depend on `L.hash`).

manmatteo committed:
Implement subtyping for polymorphic records

The `psi` function in `records.ml` implements precisely the algorithm in Castagna & Peyrot, OOPSLA2025, in place of the previous implementation that translated records to tuples. This is no longer possible since, in addition to row variables, polymorphism at the level of fields is handled with Field Type variables (which are distinct from Type Variables). Field Types are implemented using a compact representation composed of two BDDs.

Tchou commented:
Hi, nice @manmatteo . Don't hesitate to assign a review to me whenever this is ready (even partly).
I think a good way forward is to continue working on your branch until the tallying is ready (tallying is much less clear to me than subtyping). This involves rebasing your branch from time to time whenever `sstt` changes. I can help or do it directly.

manmatteo commented:
Yes I am keeping the branch updated. Maybe we could meet to start the work on tallying!

Commit:
added some test

manmatteo committed:
Make Field Variables available outside Records

manmatteo committed:
Make tallying more modular to accomodate for different variable kinds

Several constraint modules can now be written, for different variable kinds (in particular those of interest are type, field and row variables). A functor creates constraint sets over variables of a uniform kind. Sets of constraints of several kinds are kept together in a record, and sets of sets of constraints are represented by a list of such records.
